docs(homeserver): Deploy guide (#503)
pubky/pubky-core: docs/DEPLOY.md

# Deploy a Pubky Homeserver

How to make a Pubky homeserver reachable from the internet. This guide assumes you have already [installed the homeserver](./INSTALL.md) and have it running.

Commands and package names assume a Debian-based system (Ubuntu, Debian, etc.), adapt as needed for other distributions.

> **Note:** This guide assumes your homeserver is running on a machine with a static public IP and that you have a domain name you control. It covers homeserver-specific setup only, not general server hardening. If you're new to running public-facing servers, look into basic Linux server security before proceeding.

## Contents

- [Open Ports](#open-ports)
- [Configure the Homeserver](#configure-the-homeserver)
- [Set Up DNS](#set-up-dns)
- [Set Up Caddy as a Reverse Proxy](#set-up-caddy-as-a-reverse-proxy)
- [Verify the Deployment](#verify-the-deployment)
  - [Check ICANN HTTPS](#check-icann-https)
  - [Find Your Public Key](#find-your-public-key)
  - [Check PKARR Record](#check-pkarr-record)
  - [Check Pubky TLS](#check-pubky-tls)

>   - [Test a Signup](#test-a-signup)

Reviewer: AI review:

Low: docs/DEPLOY.md:20 (docs/DEPLOY.md#L20) links to #test-a-signup, but that section does not exist. Add the promised end-to-end signup test or remove the entry.

The homeserver exposes two endpoints: Pubky TLS on port 6287 and a plain HTTP endpoint on port 6286. We'll place [Caddy](https://caddyserver.com/) in front of the HTTP endpoint to serve it over HTTPS on port 443, it will manage the TLS certificates for us.

Open these three ports for inbound traffic:

| Port | Purpose |
| --- | --- |
| 80 | HTTP — [Caddy](#set-up-caddy-as-a-reverse-proxy) needs this for automatic TLS certificate provisioning. |
| 443 | HTTPS — serves your domain via [Caddy](#set-up-caddy-as-a-reverse-proxy) with a TLS certificate. |
| 6287 | Pubky TLS — direct Pubky protocol connections (no certificate needed). |

How you open these depends on your setup: a cloud provider's security group, a router's port-forwarding rules, or a host firewall.

Do **not** expose these ports:

| Port | Purpose |
| --- | --- |
| 6286 | ICANN HTTP — Caddy proxies to this internally. |
| 6288 | Admin API — sensitive operations, keep behind authentication if exposed. |
| 6289 | Metrics — internal monitoring only. |

## Configure the Homeserver

Edit `~/.pubky/config.toml` with the following settings:

```toml
[drive]
# Listen on all interfaces so Pubky TLS is reachable from the internet

> pubky_listen_socket = "0.0.0.0:6287"

Reviewer: To make the guide consistent, should the Pubky TLS socket also be proxied through Caddy? I think having it behind the proxy is better than exposing the homeserver directly. This allows to use Caddy functions for Pubky TLS too

Author: Since Caddy cant understand Pubky TLS it would be a raw TCP proxy, so it couldnt apply any of its http-level features. All we'd really get is connection-level metrics and IP-based rate limiting, which we control at the HS level anyway. 

We would require `layer4` module to support raw tcp proxying which requires a custom build of Caddy, eg:

```
# Install xcaddy
go install github.com/caddyserver/xcaddy/cmd/xcaddy@latest

# Build Caddy with the layer4 module
xcaddy build --with github.com/mholt/caddy-l4
```

I dont think its worth it.

# icann_listen_socket defaults to 127.0.0.1:6286 — no change needed.

[pkdns]
# The public-facing IP of this machine. Published to the DHT so that
# Pubky TLS clients can connect directly to the homeserver.
public_ip = "your_ip"

# Your domain name for HTTPS browser access (via the reverse proxy).
icann_domain = "your_domain.com"
```

Replace `your_ip` with the public IP of the machine running the homeserver. You can find it with `curl -4 ifconfig.me`.

> > **Important:** Ensure your server has a **static (reserved) public IP**. On most cloud providers the default external IP is ephemeral. If your ip changes then both your DNS `A` record and the PKARR record (which embeds `public_ip`) will silently point at a dead address.

Reviewer: Not sure if most are emphemeral. I would drop this sentence. Also doesn't add anything to the guide IMO.

```suggestion
> **Important:** Ensure your server has a **static (reserved) public IP**. If your ip changes then both your DNS `A` record and the PKARR record (which embeds `public_ip`) will silently point at a dead address.
```

```

These should return your server's public IP(s).

## Set Up Caddy as a Reverse Proxy

The homeserver's ICANN HTTP endpoint speaks plain HTTP. To serve it securely over HTTPS with a valid TLS certificate, you need a reverse proxy in front of it. We'll use [Caddy](https://caddyserver.com/) as its quick to setup and automatically provisions and renews TLS certificates from Let's Encrypt with zero configuration.

> **Prerequisites:** [DNS propagation](#set-up-dns) must be complete before this step, and [port 80 must be open](#open-ports). Caddy uses port 80 to complete the ACME challenge when obtaining a TLS certificate, if either DNS or port 80 are not ready then certificate provisioning will fail.

Install Caddy (Debian/Ubuntu):

```bash
sudo apt install -y debian-keyring debian-archive-keyring apt-transport-https curl
curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/gpg.key' | sudo gpg --dearmor -o /usr/share/keyrings/caddy-stable-archive-keyring.gpg
curl -1sLf 'https://dl.cloudsmith.io/public/caddy/stable/debian.deb.txt' | sudo tee /etc/apt/sources.list.d/caddy-stable.list
sudo apt update && sudo apt install -y caddy
```

> For other install methods, see the [Caddy install docs](https://caddyserver.com/docs/install).

Reviewer: I am always hesitant to have such details install instructions in a guide. As soon as Caddy changes something, this is outdated.

Isn't it enough to link to https://caddyserver.com/docs/install#debian-ubuntu-raspbian?

```
your_domain.com {
    reverse_proxy 127.0.0.1:6286
}
```

Reload Caddy:

```bash
sudo systemctl reload caddy
```

Check the Caddy logs — look for "certificate obtained" to confirm success, or any ACME errors:

```bash
journalctl -u caddy --no-pager | tail -50
```

> **Tip:** The most common cause of ACME certificate errors is port 80 not being reachable from the internet. If you see an ACME error in the logs, double-check your cloud firewall rules (e.g. GCP, AWS security groups) and any host-level firewall. If `dig` already returns your IP then DNS is not the issue.

> **Important:** The domain in the Caddyfile must exactly match both your DNS A-record name and the `icann_domain` value in `config.toml`. A mismatch between any of these is a common source of silent failures.

## Verify the Deployment

### Check ICANN HTTPS

From your local machine, verify that HTTP redirects to HTTPS and the homeserver responds:

```bash
# Should return 308 Permanent Redirect
curl -I http://your_domain.com

# Should return 200
curl -I https://your_domain.com
```

### Find Your Public Key

The remaining checks need your homeserver's public key. Retrieve it via the admin API:

```bash

> curl -s "http://127.0.0.1:6288/info" -H "X-Admin-Password: admin" | grep public_key  # use your configured password. Default is "admin".

Reviewer: AI review: 

Low: docs/DEPLOY.md:156 (docs/DEPLOY.md#L156) uses grep public_key, but /info returns compact JSON, so this prints the entire response rather than the key. Use jq -r '.public_key' and clarify that the command runs on the homeserver because port 6288 binds to localhost.

## Contents

- [Choose Your Setup](#choose-your-setup)
- [Open Ports](#open-ports)
- [Configure the Homeserver](#configure-the-homeserver)
- [Set Up HTTPS](#set-up-https)
  - [With a Domain](#with-a-domain)
  - [IP Address Only](#ip-address-only)
- [Verify the Deployment](#verify-the-deployment)
  - [Check ICANN HTTPS](#check-icann-https)
  - [Find Your Public Key](#find-your-public-key)
  - [Check PKARR Record](#check-pkarr-record)
  - [Check Internal Ports Are Not Exposed](#check-internal-ports-are-not-exposed)
  - [Check Pubky TLS](#check-pubky-tls)
- [Production Notes](#production-notes)
- [Troubleshooting](#troubleshooting)

## Choose Your Setup

There are two ways to expose your homeserver over HTTPS:

- **With a Domain** — requires a domain name and DNS setup. Standard 90-day Let's Encrypt certificates.
- **IP Address Only** — no domain needed. Uses short-lived (~6-day) Let's Encrypt certificates, so the server must stay healthy for renewals.

This guide covers both options.

## Open Ports

The homeserver speaks two protocols. Both endpoints serve the same data, the difference is how clients find and connect to your server:

> **Pubky TLS** is the native protocol: clients resolve your homeserver's public key on the DHT and connect directly on port 6287, bypassing centralized DNS.

Reviewer: bypassing centralized DNS is not the differentiating quality here. It's bypassing any certificate authority.
```suggestion
**Pubky TLS** is the native protocol: clients resolve your homeserver's public key on the DHT and connect directly on port 6287, bypassing any certificate authority.
```

> **HTTPS** is the standard web protocol: browsers and the browser-based SDK use this on port 443, which requires a TLS certificate for your domain or IP address. 

Reviewer: Both endpoints do HTTPS. The difference is cert authority vs no-cert authority. So calling it "HTTPS" feels wrong here.

Here is a chat with alternative names: https://chatgpt.com/share/6a5f2fd4-8820-83eb-9fd9-b00fbb60ddf2

Author: hmm i agree that theyre both https so this could be misleading, that name is useful though since everyone knows what it means. Instead, Ive attempted to communicate the idea of "this is the one youre used to" whilst also being technically correct but with minimal jargon. What do you think?

## Choose Your Setup

There are two ways to expose your homeserver over HTTPS:

- **With a Domain** — requires a domain name and DNS setup. Standard 90-day Let's Encrypt certificates.
- **IP Address Only** — no domain needed. Uses short-lived (~6-day) Let's Encrypt certificates, so the server must stay healthy for renewals.

This guide covers both options.

## Open Ports

The homeserver exposes two endpoints: Pubky TLS on port 6287 and a plain HTTP endpoint on port 6286. The HTTP endpoint does not handle TLS itself, so we'll place [Caddy](https://caddyserver.com/) in front of it to serve HTTPS on port 443; Caddy manages the TLS certificates automatically.

Open these three ports for inbound traffic:

| Port | Purpose |
| --- | --- |
| 80 | HTTP — [Caddy](#set-up-https) needs this for automatic TLS certificate provisioning. |
| 443 | HTTPS — serves your homeserver via [Caddy](#set-up-https). Also used for TLS-ALPN-01 certificate challenges. |
| 6287 | Pubky TLS — direct Pubky protocol connections (no certificate needed). |

How you open these depends on your setup: a cloud provider's security group, a router's port-forwarding rules, or a host firewall.

> Do **not** expose these ports:

Reviewer: ```suggestion
Do **not** expose these ports directly to the internet:
```

> **Warning:** The guide below changes `pubky_listen_socket` to `0.0.0.0:6287` so Pubky TLS is reachable externally. Do **not** apply the same `0.0.0.0` bind to `icann_listen_socket`, `admin.listen_socket`, or `metrics.listen_socket` — those must stay on `127.0.0.1` to avoid exposing internal APIs to the internet.

## Configure the Homeserver

Edit `~/.pubky/config.toml` with the following settings:

```toml
[drive]
# Listen on all interfaces so Pubky TLS is reachable from the internet
pubky_listen_socket = "0.0.0.0:6287"

# icann_listen_socket defaults to 127.0.0.1:6286 — no change needed.

[pkdns]
# The public-facing IP of this machine. Published to the DHT so that
# Pubky TLS clients can connect directly to the homeserver.
public_ip = "YOUR_IP"

# With a Domain: your domain name for HTTPS browser access.
# IP Address Only: set this to your IP address (same as public_ip).
icann_domain = "YOUR_DOMAIN_OR_IP"
```

Replace `YOUR_IP` with the public IP of the machine running the homeserver. You can find it with `curl -4 ifconfig.me`.

Replace `YOUR_DOMAIN_OR_IP` with your domain name or public IP, depending on [which setup you chose](#choose-your-setup).

> For the full list of settings (including `public_pubky_tls_port` and `public_icann_http_port` for non-standard port setups), see [`pubky-homeserver/config.sample.toml`](../pubky-homeserver/config.sample.toml).

Reviewer: This is not important in this context IMO. Remove?

Author: I think it is - the thinking is that someone not using the standard default port setup should at least be pointed to where they can learn about custom configurations.

Reviewer: Should we use demphasize it then with a `> ℹ️`?

Author: sure

> **Important:** Ensure your server has a **static (reserved) public IP**. If your IP changes then the PKARR record (which embeds `public_ip`), your Caddy configuration, and any DNS records will silently point at a dead address.

> Restart the homeserver after editing `config.toml` for the changes to take effect. If you have set it up as a background service, see [systemd Service](./INSTALL.md#systemd-service) in the install guide.

Reviewer: Changing the config.toml made me realize that it would be very convenient to have a `homeservercli config-check` command to see if the config file is valid.

Author: Yes! also some convenience functions for setting common vars, eg `homeservercli config set-database-connection-string`

Reviewer: We could. This would need to be made smart though so the user is aware what he overrides.
A flexible config set could look like this:

```
# homeservercli config set %title.key% %value%

$ homeservercli config set general.database_url "postgres://postgres:postgres@localhost:5432/pubky_homeserver"

Set "general.database_url "
from "postgres://postgres:postgres@localhost:5432/old_db"
to "postgres://postgres:postgres@localhost:5432/pubky_homeserver"
Confirm? [Yn]
```

```bash
# Use your configured admin password. Default is "admin".
curl -s "http://127.0.0.1:6288/info" -H "X-Admin-Password: admin"
```

The response is JSON — look for the `public_key` field.

### Check PKARR Record

Look up your public key on [pkdns.net](https://pkdns.net/). The record should contain your server's public IP and your domain or IP address (from `icann_domain`).

For a more thorough check, resolve the record from the DHT directly using the `resolve` example from the [pkarr](https://github.com/pubky/pkarr) repository:

```bash
cargo run --example resolve <homeserver-public-key>
```

This performs a cold lookup, a cached lookup, and a network-only lookup, printing the resolved DNS records and timings for each. Verify that the output contains an `A` record with your server's public IP and `HTTPS` (SVCB) records — one for the Pubky TLS port and one pointing to your domain or IP address.

### Check Pubky TLS

From a separate machine with the [pkarr](https://github.com/pubky/pkarr) repository cloned:

```bash
cargo run --features=reqwest-builder --example http-get https://<homeserver-public-key>
```

> A successful response prints `Pubky Homeserver`.

Reviewer: Not part of this guide but we need to make these checks more convenient. For example:
The `homeservercli` should have a check command that does this automatically.

### Check Internal Ports Are Not Exposed

From your local machine (not the server), verify that internal ports are **not** reachable. Replace `YOUR_HOST` with your domain or IP address:

```bash
# These should all time out or refuse the connection
curl --connect-timeout 5 http://YOUR_HOST:6286
curl --connect-timeout 5 http://YOUR_HOST:6288/info
curl --connect-timeout 5 http://YOUR_HOST:6289
```

> If any of these respond, your firewall is misconfigured — go back to [Open Ports](#open-ports).

Reviewer: Not needed. This is too much for a guide and should be general knowledge.

## Production Notes

- Back up the homeserver's state regularly:
  - The keypair at `~/.pubky/secret` — this is the homeserver's identity. If lost, the server cannot be recovered.
  - User data — by default stored in `~/.pubky/data/files` (depends on `storage.type`).
  - The PostgreSQL database.
- Change the default admin password in `[admin].admin_password`.

> - **IP Address Only users:** Shortlived certificates expire in ~6 days. Caddy renews them automatically, but if renewal fails for more than a few days your HTTPS endpoint will go down. Monitor your Caddy logs and ensure port 80 stays reachable.

Reviewer: No need to monitor the Caddy logs IMO

```suggestion
- **IP Address Only users:** Shortlived certificates expire in ~6 days. Caddy renews them automatically, but if renewal fails for more than a few days your HTTPS endpoint will go down. Ensure port 80 stays reachable.
```
